### DISCUSSION TITLE
Enable third-party deserialization in parse.py (until a better solution appears in Pydantic)

### DISCUSSION THREAD

**Discussion by NowanIlfideme** [pydantic/pydantic#3025] [Feature Request]
Pydantic currently only supports parsing JSON (and pickle) files.

I'm well aware of this comment that `parse.py` is intended to be removed: https://github.com/samuelcolvin/pydantic/issues/136#issuecomment-370232244

3 years have passed and [it's still here](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/parse.py).
However, this year @samuelcolvin mentioned having a potential solution: https://github.com/samuelcolvin/pydantic/issues/2335#issuecomment-776133074

As I understand, the main concern is having yet more code in Pydantic that needs to be tested and maintained. I really understand this concern - smaller scope, higher flexibility for end users, etc.

The current approach is to parse on a case-by-case:

```python
dict_like = xxx.loads(data)
model = Model.parse_obj(**dict_like)
```

However, there are multiple cases where the developer would like to support various configurations. I've written "to/from YAML" helpers several times before writing `pydantic_yaml`. XML and TOML support was another thing I saw. I can imagine some projects might even need an INI config parser (e.g. to support legacy configs). And being able to specify `model = Model.parse_raw(raw_data, content_type="application/yaml")` based on some received MIME type, for example, makes code significantly more readable.

## Proposal

Make a minor change in `parse.py` to allow user code (libraries or custom) to register their own "Protocol" for loading from str or bytes. This can keep existing logic as needed (for backward compatibility), but instead of erroring will try additional "protocols" that have been registered. This registration can happen after Pydantic has loaded, e.g. by user code or by another library being imported. When a better solution is implemented by Pydantic, `parse.py` can be deprecated completely, and the dependent packages will have to update to the better solution.

### The Alternative

Make no changes in Pydantic. External libraries (such as `pydantic_yaml`) will have to either implement their own classes for loading (as I currently do) or monkey-patch `parse.py` (which is bad practice) until a "good solution" is implemented by Pydantic.

## Other thoughts

A similar thing could be used for dumping, but the API here is significantly different. Here I think it makes sense to just use mixins that supply dumpers (such as `model.yaml()`) rather than having an explicit entry point. No changes are needed here, from Pydantic!

The main obstacle to backward-compatible changes is the so-called `Protocol` class defined [here](https://github.com/samuelcolvin/pydantic/blob/5ccbdcb5904f35834300b01432a665c75dc02296/pydantic/parse.py#L10) (not to be confused with `typing.Procotol`). You can't change an enum after definition - the mechanism for "protocol" registration would be different. Is this class considered an internal implementation detail, or are external users also using it? It's exported in the main namespace... My current thinking is to leave it as-is, but allow `Union[Protocol, str]` in functions instead.

I would be willing to do the PR, however it only makes sense to work on it if @samuelcolvin agrees with the above.
Alternatively, it would great to know about preliminary thinking in the "remove direct parsing support" direction for Pydantic.

**Comment by NowanIlfideme** [comment]
I ended up making this pull request, since the change seems pretty trivial... https://github.com/samuelcolvin/pydantic/pull/3026

**Comment by PrettyWood** [comment]
Hi @NowanIlfideme 
I completely agree with you and really want the v2 of pydantic to be more flexible to allow us register our own custom serialization/deserialization but also types with ease...
I'll take some time next week to try out the API of the protocol in your PR

**Reply by NowanIlfideme** [reply]
By the way, I just re-read this and realized you said "register custom types"... would this potentially be for including custom objects (e.g. Dataframes) in models, with some similar system to register how to load/dump the classes from a file format? I'd love to be a fly on the wall for that discussion. 😉

**Comment by zrothberg** [comment]
So I ran into similar problems with integrating dynamodb into pydantic. I am wondering if pydanitc exposing a more descriptive ModelField definitions would allow this problem to be easily solvable with third party parsers/encoders. I think expanding on the values inside of __fields__ along with maybe making them accessible in a way without accessing the dict (maybe as like a class level field) would go a long way to making easily created third party parsers and encoders. I think this lib does something along those lines allowing you to access them in the class level object https://art049.github.io/odmantic/engine/#fetch-a-single-instance.

Not sure if this is a subject that has come up before.
